src/blocked.c

```diff
@@ -0,0 +1,253 @@
+        return;
+    }
+
+    sc_list_del(&redis_raft.blocked_command_list, &blocked->blocked_list);
+}
+
+void freeBlockedCommand(BlockedCommand *bc)
+{
+    if (bc->command) {
+        RedisModule_Free(bc->command);
+        bc->command = NULL;
+    }
+    if (bc->data) {
+        RedisModule_Free(bc->data);
+        bc->data = NULL;
+    }
+    if (bc->reply) {
+        RedisModule_FreeCallReply(bc->reply);
+        bc->reply = NULL;
+    }
+    RedisModule_Free(bc);
+}
+
+BlockedCommand *getBlockedCommand(raft_index_t idx)
+{
+    return RedisModule_DictGetC(redis_raft.blocked_command_dict, &idx, sizeof(idx), NULL);
+}
+
+void clearAllBlockCommands()
+{
+    struct sc_list *elem;
+    while ((elem = sc_list_pop_head(&redis_raft.blocked_command_list)) != NULL) {
+        BlockedCommand *bc = sc_list_entry(elem, BlockedCommand, blocked_list);
+        if (RedisModule_CallReplyPromiseAbort(bc->reply, NULL) != REDISMODULE_OK) {
+            /* shouldn't happen with normal redis commands */
+            LOG_WARNING("timeoutBlockedCommand: failed to abort %lu", bc->idx);
+            return;
+        }
+        freeBlockedCommand(bc);
+    }
+    if (redis_raft.blocked_command_dict != NULL) {
+        RedisModule_FreeDict(redis_raft.ctx, redis_raft.blocked_command_dict);
+    }
+    redis_raft.blocked_command_dict = RedisModule_CreateDict(redis_raft.ctx);
+}
+
+void blockedCommandsSave(RedisModuleIO *rdb)
+{
+    RaftSnapshotInfo *info = &redis_raft.snapshot_info;
+    struct sc_list *it;
+    int count = 0;
+
+    sc_list_foreach (&redis_raft.blocked_command_list, it) {
+        BlockedCommand *bc = sc_list_entry(it, BlockedCommand, blocked_list);
+        if (bc->idx <= info->last_applied_idx) {
+            count++;
+        } else {
+            break;
+        }
+    }
+
+    RedisModule_SaveUnsigned(rdb, count);
+    sc_list_foreach (&redis_raft.blocked_command_list, it) {
+        BlockedCommand *bc = sc_list_entry(it, BlockedCommand, blocked_list);
+        if (bc->idx <= info->last_applied_idx) {
+            RedisModule_SaveUnsigned(rdb, bc->idx);
+            RedisModule_SaveUnsigned(rdb, bc->session);
+            RedisModule_SaveStringBuffer(rdb, bc->data, bc->data_len);
+        } else {
+            break;
+        }
+    }
+}
+
+void blockedCommandsLoad(RedisModuleIO *rdb)
+{
+    RedisRaftCtx *rr = &redis_raft;
+
+    clearAllBlockCommands();
+
+    size_t command_count = RedisModule_LoadUnsigned(rdb);
+    for (size_t i = 0; i < command_count; i++) {
+        raft_index_t idx = RedisModule_LoadUnsigned(rdb);
+        raft_session_t session = RedisModule_LoadUnsigned(rdb);
+        size_t data_len;
+        char *data = RedisModule_LoadStringBuffer(rdb, &data_len);
+
+        /* execute blocking command again */
+        RaftRedisCommandArray tmp = {0};
+        if (RaftRedisCommandArrayDeserialize(&tmp, data, data_len) != RR_OK) {
+            PANIC("Invalid Raft entry");
+        }
+        tmp.client_id = session;
```

Comment:
```suggestion
        RaftRedisCommandArray tmp = {
            .client_id = session,
        };
        if (RaftRedisCommandArrayDeserialize(&tmp, data, data_len) != RR_OK) {
            PANIC("Invalid Raft entry");
        }
```
You can move it into initializer if you want

Reply:
ok